code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/process/ServiceDiscoveryInstanceStartPostListener.java

```diff
@@ -48,31 +50,34 @@
     @Override
     public HandlerResult handle(ProcessState state, ProcessInstance process) {
         Instance instance = (Instance) state.getResource();
-        Long serviceId = null;
         if (lbInstanceService.isLbInstance(instance)) {
             LoadBalancer lb = lbInstanceService.getLoadBalancerForInstance(instance);
-            serviceId = lb.getServiceId();
+            Long serviceId = lb.getServiceId();
+            if (serviceId != null) {
+                // for the lb instance, service map gets created only at this point
+                // as the instance gets created in generic manner by AgentBuilder factory, and I avoided to make service
+                // specific modifications there
+                ServiceExposeMap instanceServiceMap = mapDao.findNonRemoved(ServiceExposeMap.class, Instance.class,
+                        instance.getId(),
+                        Service.class, serviceId);
+
+                if (instanceServiceMap == null) {
+                    instanceServiceMap = resourceDao.createAndSchedule(ServiceExposeMap.class,
+                            SERVICE_EXPOSE_MAP.INSTANCE_ID,
+                            instance.getId(), SERVICE_EXPOSE_MAP.SERVICE_ID, serviceId);
+                }
+            }
         } else {
-            Integer requestedServiceId = (Integer)DataUtils.getFields(instance).get(
-                    ServiceDiscoveryConstants.FIELD_SERVICE_ID);
-            if (requestedServiceId != null) {
-                serviceId = requestedServiceId.longValue();
+            // for regular, non lb instance, the instance->service map gets created as a part of the instance creation
+            // (within the same transaction), so it should exist at this point
+            List<? extends ServiceExposeMap> instanceServiceMap = mapDao.findNonRemoved(ServiceExposeMap.class,
+                    Instance.class,
+                    instance.getId());
+            if (instanceServiceMap.isEmpty()) {
+                return null;
```

Comment:
Just because I know I'm going to forget in the future, can we add a comment here.  Something like: "not service instance"?